### ISSUE TITLE
New example: Subscriptions

### ISSUE THREAD

**Issue by newswim** [dwyl/learn-elm#74]
@nelsonic asked (#12) if someone would be up for adding a _websockets_ example.

Since `Subscriptions` can use be used to listen for events via an always-on connection, maybe we should come on with a subscription example that uses WebSockets.. ?

In his blog post, "[Farewell To FRP]", the author (of Elm) states that _signals_ were a major hurdle for newcomers. As of 0.17, Signals are out, and a more generic form of subscribing to events is _in_. 💯 

> Subscriptions let components sit around and wait for messages while library code handles a bunch of tricky resource management stuff behind the scenes.

> In the end, it was possible to remove signals because Elm has been moving towards an explicit emphasis on **concurrency** for quite some time now.

### Nota

"[upgrade plan]" for 0.16 -> 0.17
[ports] also work with subscriptions now!

**Subscription Examples** : [One], [Two-hoo-hoo-hooo], [Three]. 🦉 

P.S. The API for Elm's [WebSocket] library is very minimal.

### TL;DR 

- [ ] Let's come up with an example using _Subscriptions!!!_ 

[One]: http://elm-lang.org/examples/websockets
[Two-hoo-hoo-hooo]: https://www.elm-tutorial.org/en/03-subs-cmds/01-subs.html
[Three]: https://ohanhi.github.io/base-for-game-elm-017.html
[WebSocket]: http://package.elm-lang.org/packages/elm-lang/websocket/latest/WebSocket
[Farewell To FRP]: http://elm-lang.org/blog/farewell-to-frp
[upgrade plan]: https://github.com/elm-lang/elm-platform/blob/master/upgrade-docs/0.17.md
[ports]: https://evancz.gitbooks.io/an-introduction-to-elm/content/interop/javascript.html

**Comment by newswim** [comment]
Fair warning, this is a speculative and very naive suggestion. If I have the time in the next few weeks, I'd like to tackle it, but I don't even know enough elm to know what I don't know yet.

**Comment by njsfield** [comment]
@newswim great suggestion! The examples so far demonstrate using Http **beginnerProgram** function, perhaps adding an example that uses the program function with subscriptions could work well if the effect subscribed to was a simple browser event.

The WebSocket subscription examples are cool, but the learner would need to have an understanding of what WebSockets are to contextualise whats going on. I created [this](https://github.com/njsfield/elm-websocket-chat
) chatroom example- but it requires a backend with a server using the [ws](https://www.npmjs.com/package/ws
) npm module so is a bit too heavy.

How about an example using a browser event like Window resizing? This [gist](https://gist.github.com/chrisbuttery/bf4d50c822df56db6013ac12b82cea4f
) provides a nice example using an earlier package version of the [Window](http://package.elm-lang.org/packages/elm-lang/window/latest/Window
) package.

Or maybe an example based on [this](https://guide.elm-lang.org/architecture/effects/time.html
) tutorial on using the [Time](http://package.elm-lang.org/packages/elm-lang/core/latest/Time
) package. This may be more suitable as Time is built into the latest core library (whereas the Window example is based on an older package to take advantage of the *resizes* function)

> I'd like to tackle it, but I don't even know enough elm to know what I don't know yet.

Am happy to help if you’ve got an idea but have questions on unknown areas :thumbsup:

**Comment by newswim** [comment]
Hey @njsfield, thanks so much for the feedback and offer to help. I'm really keen on learning Elm in more depth, and this repo is where I always others who are learning.

I updated the [space-invaders](https://github.com/newswim/elm-space-invaders) example to use 0.18, which also uses the `Time` package. But as you rightly point out, there is some familiarity with WebSockets and their protocol to really understand what subscriptions are doing behind the scenes. As an architectural decision, i think abstracting the wiring and giving users a first-class way of using and subscribing in real-time is great! I myself started learning web-development with Meteor, which does something similar, and has a nice fall-back plan if WS's aren't available.

I love the example using `Window` -- maybe we should write something brief explaining what WebSockets are? Here's the [docs](https://github.com/meteor/meteor/blob/devel/packages/ddp/DDP.md) for Meteor's DDP package.
